[leroytan] iP

leroytan opened the pull request:
### Light

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Light frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/leroytan/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is FREE!

Features:

- [x]  Managing tasks
- [ ]  Managing deadlines
- [ ]  Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

leroytan committed:
Level-0

leroytan committed:
Level-1

leroytan committed:
Level-2

leroytan committed:
Level-3

leroytan committed:
Level-4

leroytan committed:
Level-5

leroytan committed:
Level-6

leroytan committed:
A-TextUiTesting

leroytan committed:
Implement saved list

The program currently lacks the functionality to persist data across
sessions, which means any information stored in the list is lost when
the program closes.

This limitation makes the application less user-friendly, as users
have to re-enter their data every time they run the program.
Persistent storage is necessary to improve usability.

Add methods to read and write to file to handle data storage and
retrieval.

The 'writeToFile()' method is used to store the current state of the
list into a file, 'arrayToNumberedString()' formats the data for
readability, and 'readFromFile()' is used to retrieve the stored data
when the program runs again. This approach ensures that data is
consistently saved and retrieved in a structured manner.

leroytan committed:
Implement saved list

The program currently lacks the functionality to persist data across
sessions, which means any information stored in the list is lost when
the program closes.

This limitation makes the application less user-friendly, as users
have to re-enter their data every time they run the program.
Persistent storage is necessary to improve usability.

Add methods to read and write to file to handle data storage and
retrieval.

The 'writeToFile()' method is used to store the current state of the
list into a file, 'arrayToNumberedString()' formats the data for
readability, and 'readFromFile()' is used to retrieve the stored data
when the program runs again. This approach ensures that data is
consistently saved and retrieved in a structured manner.

leroytan committed:
Merge branch 'branch-Level-7'

leroytan committed:
Implement Datetime handling

The chatbot currently treats dates and times as plain strings,
limiting its ability to manage tasks effectively.

Proper date and time handling is crucial for managing deadlines and
events, making the chatbot more useful and user-friendly.

leroytan committed:
Merge branch 'branch-Level-8'

leroytan committed:
A-MoreOOP (Changed code to obey OOP)

Refactoring the code to adhere to Object-Oriented Programming (OOP)
principles will improve maintainability and scalability,
making the code more modular and easier to extend.

Extract the code into separate classes:
Ui: Handles user interactions.
Storage: Manages loading and saving tasks from/to a file.
Parser: Interprets user commands.
TaskList: Contains the task list and provides operations for
adding and deleting tasks.
Command: Represents the instructions the program can receive

leroytan committed:
A-Packages (Put classes into packages)

All classes in the Light program are currently in a single namespace,
making it harder to manage and scale as the program grows.

Organizing the classes into packages improves code organization,
readability, and maintainability. It also allows for easier management
of the growing codebase as more classes are added.
Grouping related classes into packages also enhances modularity and
helps maintain a clean project structure.

leroytan committed:
Merge tag 'A-Packages'

leroytan committed:
Merge branch 'add-gradle-support' of https://github.com/leroytan/ip

leroytan committed:
Add Gradle support

leroytan committed:
A-JUnit

leroytan committed:
A-Jar

leroytan committed:
A-JavaDoc

leroytan committed:
Adhere to A-CodingStandard

leroytan committed:
Add Level-9(Find)

leroytan committed:
Merge tag 'A-JavaDoc'

leroytan committed:
Merge tag 'A-CodingStandard'

# Conflicts:
#	src/main/java/Light.java
#	src/main/java/command/AddCommand.java
#	src/main/java/command/DeleteCommand.java
#	src/main/java/command/ListCommand.java
#	src/main/java/command/MarkCommand.java
#	src/main/java/components/Storage.java
#	src/main/java/components/Ui.java
#	src/main/java/exceptions/LightException.java

leroytan committed:
Merge tag 'Level-9'

# Conflicts:
#	src/main/java/components/Parser.java

wujiaweijoanna commented on src/main/java/components/Storage.java:
```diff
@@ -0,0 +1,130 @@
+
+            while (scanner.hasNextLine()) {
+                String data = scanner.nextLine();
+                Boolean marked = data.charAt(7) == 'X';
```
Maybe can change the boolean variable name to isMarked?

wujiaweijoanna commented on src/main/java/Light.java:
```diff
@@ -0,0 +1,60 @@
+        while (!isExit) {
+            try {
+                String fullCommand = ui.readCommand();
+                Command c = Parser.parse(fullCommand);
```
Maybe can change c to command instead?

wujiaweijoanna commented on src/main/java/command/DeleteCommand.java:
```diff
@@ -0,0 +1,39 @@
+ * Represents a command to delete a task from the task list.
+ */
+public class DeleteCommand extends Command{
+    private int taskNumber;
```
All variable and method name in the pr follows the naming convention 👍

wujiaweijoanna commented on src/main/java/task/TaskList.java:
```diff
@@ -0,0 +1,35 @@
+     * @return A TaskList containing tasks that match the keyword.
+     */
+    public TaskList findTasks(String keyword) {
+        ArrayList<Task> foundTasks = new ArrayList<>();
```
Plural form is used on names representing a collection of objects 👍

wujiaweijoanna reviewed (commented):
Looks good to me. 🎉
Some minor adjustments suggested.

PrishaVP commented on src/main/java/command/AddCommand.java:
```diff
@@ -0,0 +1,39 @@
+package command;
```
not sure whether you have to do this but the naming convention for packages said if its for a school project then the root name of the package should be your project name so maybe need to name the package starting with ip or light? I'm not fully sure though

PrishaVP commented on src/main/java/components/Ui.java:
```diff
@@ -0,0 +1,71 @@
+    /**
+     * Closes the user interface.
+     */
+    public void closeUI() {
```
maybe can rename the method to closeUi (i at the end as lowecase) in order to meet this code convention: Abbreviations and acronyms should not be uppercase when used as a part of a name.

PrishaVP reviewed (commented):
Code conventions lgtm except for a few minor things

leroytan committed:
Add GUI (Level-10)

leroytan committed:
Merge tag 'Level-10'

leroytan committed:
Add Assertions

The program does not currently include assert statements to verify key
assumptions at runtime.

Using assertions ensures that critical assumptions are valid, helping to
catch potential errors early in development.

Add assert statements at important points in the code to document and
validate assumptions.

Assertions act as a safeguard, only running during development, without
impacting performance in production when disabled.

leroytan committed:
Improve CodeQuality

Command represented by strings, poor naming of variables affects code
readability

Replacement of command strings by enums, restructuring of code allows
short and readable code

leroytan committed:
Merge tag 'A-Assertions'

leroytan committed:
Merge tag 'A-CodeQuality'

# Conflicts:
#	src/main/java/command/ListCommand.java
#	src/main/java/components/Light.java
#	src/main/java/components/Storage.java

leroytan committed:
Fix bug where the brackets are printed twice

leroytan committed:
Merge branch 'master' into C-Undo

leroytan committed:
Add Undo and Redo commands

The program currently lacks the ability to undo or redo actions, making
it difficult for users to correct mistakes or revert changes.

Undo and redo functions improve the user experience by allowing users
to reverse unintended actions and restore previous states.

Implement an undo/redo feature by tracking changes to the task list
and allowing users to step backward or forward through their action
history.

This approach ensures actions can be easily reverted or reapplied,
improving flexibility and usability.

The implementation stores previous states to support multiple undo/redo
steps.

leroytan committed:
Merge branch 'C-Undo'

leroytan committed:
Improve gui

Tweak the GUI to match the asymmetric nature of the conversation: As the
 conversation is between the user and the app (not between two humans),
it is asymmetric in nature. Accordingly, it makes sense not to display
both sides of the conversion in the same visual format.

Add text bubble, add background, edit visuals

leroytan committed:
A-UserGuide

Update the given docs/README.md

leroytan committed:
Update product image on README

Product image is not being shown on README.

Fix a bug where the product image on README was not shown
by changing the filepath.

leroytan committed:
Remove dead comments

Dead comments are comments used during development or testing
and are no longer utilised.

Dead comments are removed from codebase to clean up and simplify code.

leroytan committed:
Update README to include find and help instruction

README is missing the find and help instruction.

Update README to include the find and help instruction with usage,
example and expected output.

leroytan committed:
Merge PR

This is a commit to fufil the requirements of the IP project, which is
to try merging a PR with my remote repo on Github.

leroytan committed:
Update find to work without keyword

Program crashes when we use find without a keyword.

Fix a bug where findcommand would crash when there is no keyword.
